### PR TITLE
Fix  run_jedi_hofx_ensemble_executable.py  related to LocalEnsembleDA

### PR DESCRIPTION
**Pull request by metdyn**
## Description

See details in
- https://github.com/GEOS-ESM/swell/issues/404

## This is a prerequisiste for
- https://github.com/GEOS-ESM/swell/pull/397

## Unresolved issue

- swell test code_tests failed to detect  polymorphism:
class RunJediHofxEnsembleExecutable(RunJediHofxExecutable):
class RunJediHofxExecutable(taskBase):

so that code_tests tries to regenerate  task_questions.yaml by removing `RunJediHofxEnsembleExecutable and RunJediLocalEnsembleDaExecutable`, because it is not smart enough. 
```
tasks/task_questions.yaml
save_geovals:
  ask_question: false
...
  tasks:
  - RunJediHofxExecutable
  - RunJediHofxEnsembleExecutable
  - RunJediLocalEnsembleDaExecutable
  type: boolean
```

This is a dilemma:
If I use my own  questions.yaml,  code_tests will fail. 
If I use the regenerated  questions.yaml,  I got the following error in the end of the text.

We have several choices:
- use tricks to by pass  code_tests, because it does not see polymorphism
- fix the bug in code_tests
- eliminate  super().execute(ensemble_members=packet_ensemble_members)  in enHX, stop using this polymorphism, and just transplant the needed code from HX to enHX. 

What is the previous experience to overcome the defects in test_codes?

│ File "/gpfsm/dnb07/projects/p234/yyu11/br_jedi_scu17/install_swell/src/swell/swell_lenda/src/swell/tasks/run_jedi_hofx_ensemb\
le_executable.py", line 141, in execute
│ │ super().execute(ensemble_members=packet_ensemble_members)

│ File "/gpfsm/dnb07/projects/p234/yyu11/br_jedi_scu17/install_swell/src/swell/swell_lenda/src/swell/tasks/run_jedi_hofx_execut\
able.py", line 42, in execute
│ │ save_geovals = self.config.save_geovals()

│ File "/gpfsm/dnb07/projects/p234/yyu11/br_jedi_scu17/install_swell/src/swell/swell_lenda/src/swell/utilities/config.py", line\
 135, in variable_not_found
│ │ self.__logger__.abort(f'In config class, trying to get variable \'{name}\' but ' +

### CODE REVIEW

**Comment by Dooruk** [comment]
@metdyn, can you please merge `develop` into this and resolve conflicts, aftrerwards I will test it. Thanks!

File: src/swell/suites/localensembleda/flow.cylc [hunk 1]
```diff
@@ -42,8 +47,19 @@
         {% if cycle_time[model_component] %}
             # Task triggers for: {{model_component}}
             # ------------------
+            # Generate satellite channel records
+            CloneGeosMksi-{{model_component}}[^] => GenerateObservingSystemRecords-{{model_component}}
+
             # Perform staging that is cycle dependent
-            CloneJedi[^] => StageJediCycle-{{model_component}}
+            #BuildJediByLinking[^]? | BuildJedi[^] => StageJediCycle-{{model_component}}
+            StageJediCycle-{{model_component}}
```
Comment (Dooruk) [inline src/swell/suites/localensembleda/flow.cylc:55]: StageJedi has a dependency here, uses some files from the linked JEDI build directory in the experiment folder. So we should maintain that.

File: src/swell/test/suite_tests/localensembleda-tier1.yaml [hunk 2]
```diff
@@ -1,19 +1,19 @@
 final_cycle_point: '2021-12-12T00:00:00Z'
 jedi_build_method: use_existing
+save_geovals: true
```
Comment (Dooruk) [inline src/swell/test/suite_tests/localensembleda-tier1.yaml:3]: I don't think this is needed for ensemble, could be set to false.

File: src/swell/test/suite_tests/localensembleda-tier1.yaml [hunk 3]
```diff
@@ -1,19 +1,19 @@
...
 model_components: ['geos_atmosphere']
 models:
   geos_atmosphere:
-    horizontal_resolution: '4'
-    npx_proc: 2
-    npy_proc: 2
+    npx_proc: 1
+    npy_proc: 1
```
Comment (Dooruk) [inline src/swell/test/suite_tests/localensembleda-tier1.yaml:8]: Particular reasons to decrease processor counts?

**Review by Dooruk — COMMENTED** [review]
Looks like there is an issue with window_type. Ensmeanvariance works with 3D but not with the 4D window type. On the contrary, enshofx requires 4D window type. I'm pretty sure we would like 4D window type for enshofx (@rtodling could correct me in today's meeting) so we just need to figure out to run Ensmeanvariance that way?

Your second comment pertains the inheritance rather than polymorphism perhaps? Child classes obtaining parent inputs. With the way `task_questions.yaml` is setup, that is not quite possible. We used inheritance originally but then abandoned it, I see the benefit of it for ensemble tasks, so we could revisit that.

Some potential strategies were discussed here:
https://github.com/GEOS-ESM/swell/issues/314

File: src/swell/suites/localensembleda/flow.cylc (hunk 1 above)
Comment (metdyn) [reply src/swell/suites/localensembleda/flow.cylc:55]: Thanks, that makes sense.  I had a miscocept that R1 will run first before the Tij group, which is incorrect.   Code corrected.

File: src/swell/test/suite_tests/localensembleda-tier1.yaml (hunk 2 above)
Comment (metdyn) [reply src/swell/test/suite_tests/localensembleda-tier1.yaml:3]: Removing this key helps to pass `swell test code_tests`.  The default in task_quetions.yaml is false.

File: src/swell/test/suite_tests/localensembleda-tier1.yaml (hunk 3 above)
Comment (metdyn) [reply src/swell/test/suite_tests/localensembleda-tier1.yaml:8]: oops, it is for simple test only and will be restore once I am done with the development.

**Comment by metdyn** [comment]
I am closing the PR to work offline to solve many issues.  Thanks, @Dooruk!

**Pull request closed without merging**